backend/push.js

```diff
@@ -35,72 +36,47 @@ const saveSubscription = (server, subscriptionId) => {
   server.pushSubscriptions[subscriptionId].subscriptions.delete(channelID)
-  saveSubscription(server, subscriptionId)
+  return saveSubscription(server, subscriptionId)
 }
 
-// Generate an UUID from a `PushSubscription'
-export const getSubscriptionId = async (subscriptionInfo: Object): Promise<string> => {
-  const textEncoder = new TextEncoder()
-  // <https://w3c.github.io/push-api/#pushsubscription-interface>
-  const endpoint = textEncoder.encode(subscriptionInfo.endpoint)
-  // <https://w3c.github.io/push-api/#pushencryptionkeyname-enumeration>
-  const p256dh = textEncoder.encode(subscriptionInfo.keys.p256dh)
-  const auth = textEncoder.encode(subscriptionInfo.keys.auth)
-
-  const canonicalForm = new ArrayBuffer(
-    8 +
-      (4 + endpoint.byteLength) + (2 + p256dh.byteLength) +
-      (2 + auth.byteLength)
-  )
-  const canonicalFormU8 = new Uint8Array(canonicalForm)
-  const canonicalFormDV = new DataView(canonicalForm)
-  let offset = 0
-  canonicalFormDV.setFloat64(
-    offset,
-    subscriptionInfo.expirationTime == null
-      ? NaN
-      : subscriptionInfo.expirationTime,
-    false
-  )
-  offset += 8
-  canonicalFormDV.setUint32(offset, endpoint.byteLength, false)
-  offset += 4
-  canonicalFormU8.set(endpoint, offset)
-  offset += endpoint.byteLength
-  canonicalFormDV.setUint16(offset, p256dh.byteLength, false)
-  offset += 2
-  canonicalFormU8.set(p256dh, offset)
-  offset += p256dh.byteLength
-  canonicalFormDV.setUint16(offset, auth.byteLength, false)
-  offset += 2
-  canonicalFormU8.set(auth, offset)
-
-  const digest = await crypto.subtle.digest('SHA-384', canonicalForm)
-  const id = Buffer.from(digest.slice(0, 16))
-  id[6] = 0x80 | (id[6] & 0x0F)
-  id[8] = 0x80 | (id[8] & 0x3F)
-
-  return [
-    id.slice(0, 4),
-    id.slice(4, 6),
-    id.slice(6, 8),
-    id.slice(8, 10),
-    id.slice(10, 16)
-  ].map((p) => p.toString('hex')).join('-')
+const removeSubscription = async (subscriptionId) => {
+  try {
+    const server = sbp('okTurtles.data/get', PUBSUB_INSTANCE)
+    const subscription = server.pushSubscriptions[subscriptionId]
+    if (subscription) {
+      delete server.pushSubscriptions[subscriptionId]
+      if (server.subscribersByChannelID) {
+        subscription.subscriptions.forEach((channelID) => {
+          server.subscribersByChannelID[channelID]?.delete(subscription)
+        })
+      }
+    } else {
+      // this can happen for example when a new subscription is added but then
+      // immediately removed because postEvent got a 401 Unauthorized when adding
+      // the new subscription. In this case removeSubscription could be later called
+      // again by the client but it's already been removed
+      console.warn(`removeSubscription: non-existent subscription '${subscriptionId}'`)
+    }
+    await deleteSubscriptionFromIndex(subscriptionId)
+    await sbp('chelonia.db/delete', `_private_webpush_${subscriptionId}`)
```

Comment:
I realise this was like this from before, but I'm wondering if maybe the order of these two operations should be reversed

Reply:
I don't know, if there's a strong reason for it that can be given I'm certainly happy to do it.

Reply:
The reason is that if the deletion fails, then it'll still be in the index.

Reply:
It doesn't 'fix' it, but it being in the index but not existing is worse than the converse, based on the assumption that keys may not be enumerable.